Include/object.h

```diff
@@ -155,6 +180,9 @@ static inline int _Py_IS_TYPE(const PyObject *ob, const PyTypeObject *type) {
 
 
 static inline void _Py_SET_REFCNT(PyObject *ob, Py_ssize_t refcnt) {
+    if (_Py_IsImmortal(ob)) {
+        return;
+    }
```

Comment:
Should this case be an error?  If I called `Py_SET_REFCNT()` on an immortal object then afterward I'd expect the refcount to match what I passed.  Maybe it should be a warning?

Reply:
Yeah, I do agree that calling `Py_SET_REFCNT ` and not having the value set is a bit unintuitive. However, the main motivation for this is as a defense mechanism for code written without the assumption that instances could be immortal. The runtime probably knows best when an instance should be immortal so if it has been marked as so, we probably should keep it as is.

For instance, even in the current runtime we have an example where `Py_SET_REFCNT` would completely get rid of the immortalization benefit had it not been for this check here. This is in the `PyModuleDef_Init` C-APi. i.e:
```
 #define PyModuleDef_HEAD_INIT { \
   PyObject_IMMORTAL_HEAD_INIT
   ...
  }

// foo refcount is immortal
static struct PyModuleDef foo = {
  PyModuleDef_HEAD_INIT
};

// foo refcount will now be 1.
PyModuleDef_Init(&foo);
```

-----------

Now, going back to your questions, if we make it an error then we'll force C-Extension users to fix their code like the following:
```
if (! _Py_IsImmortal(obj) {
  Py_SET_REFCNT(obj, 1);
}
```
In which case, it might make sense to just keep it within the function? 

If we make it into a warning, then we'll have no way of distinguishing between legitimate use cases where users want to actively overwrite the immortalization of an instance. This would make the application a bit annoying since it'll always warn on a legitimate use case.

Perhaps, the best option would be to just introduce a new C-API: `Py_Mortalize` so that users who actively want to override the immortalization can do:
```
Py_Mortalize(obj):
Py_SET_REFCNT(obj, 1);
```

What do you think? Otherwise, I'm wouldn't be too opposed to just dropping this and let `Py_SET_REFCNT` keep its current behavior and just going through and fixing all use cases. I just think that this defense mechanism would be a neat way to keep immortality even with code written under older assumptions.

Reply:
I think it's fine the way you have it.  You're right about everything. 🙂  It would probably be best to just add a note in the docs.

Also, I don't think we should worry about `Py_Mortalize()` until it's clear someone needs it.

Reply:
>  It would probably be best to just add a note in the docs.

For sure! I will update the docs once I get a thumbs up on the PR!

Reply:
If there is a Py_Mortalize(obj) (or maybe _Py_Mortalize(obj)) shouldn't it also take a reference count?  Reference count should certainly not be undefined on a mortal object.  I suppose you could have a convention that it starts with 1, but that seems like a situation where Explicit Is Better Than Implicit.